ext_mongodb.php

```diff
@@ -718,17 +775,82 @@ function __debugInfo() : array;
 
 final class Javascript implements Type, \Serializable
 {
-	use DenySerialization;
-
 	private $code;
 
-	public function __construct(string $code, private ?mixed $scope = NULL)
+	static private function checkArray(array $state)
+	{
+		if (
+			!array_key_exists( 'code', $state ) || !is_string( $state['code'] )
+		) {
+			throw new \MongoDB\Driver\Exception\InvalidArgumentException( "MongoDB\BSON\Javascript initialization requires \"code\" string field" );
+		}
+
+		if ( array_key_exists( 'scope', $state ) && ( $state['scope'] !== NULL ) )
+		{
+			if ( !is_array( $state['scope'] ) && !is_object( $state['scope'] ) )
+			{
+				$valueType = gettype( $state['scope'] );
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException( "Expected scope to be an array or object, {$valueType} given" );
+			}
+		}
+	}
+
+	public function serialize() : string
+	{
+		$s = [];
+		$s['code'] = $this->code;
+		if ( isset( $this->scope ) )
+		{
+			$s['scope'] = $this->scope;
+		}
+		return serialize( $s );
+	}
+
+	public function unserialize(mixed $serialized) : void
+	{
+		$unserialized = unserialize( $serialized );
+		self::checkArray( $unserialized );
+		if ( array_key_exists( 'scope', $unserialized ) )
+		{
+			$this->__construct( $unserialized['code'], $unserialized['scope'] );
+		}
+		else
+		{
+			$this->__construct( $unserialized['code'] );
+		}
+	}
+
+	public function __construct(string $code, ?mixed $scope = NULL)
 	{
 		if ( strstr( $code, "\0" ) !== false )
 		{
 			throw new \MongoDB\Driver\Exception\InvalidArgumentException( "Code cannot contain null bytes" );
 		}
+		if ( $scope !== NULL && ! ( is_object( $scope ) || is_array( $scope ) ) )
+		{
+			$valueType = gettype( $scope );
+			throw new \MongoDB\Driver\Exception\InvalidArgumentException( "Expected scope to be an array or object, {$valueType} given" );
+		}
+
 		$this->code = $code;
+		if ( $scope !== NULL )
+		{
+			$this->scope = (object) $scope;
+		}
+	}
+
+	static public function __set_state(array $state)
+	{
+		self::checkArray( $state );
+
+		if ( array_key_exists( 'scope', $state ) )
```

Comment:
Can you also use `??` here and always invoke the constructor with both arguments?